Reactify validator form

jason-crowley opened the pull request:
Migrate the validator form from embedded Ruby to React to begin the transition to more dynamic form validation.

okeefm commented on src/components/ResourceForm.tsx:
```diff
@@ -0,0 +1,20 @@
+import React, { useState } from 'react';
+
+export function ResourceForm() {
```
Might be worth renaming these from `<x>Form` to `x<Element>` or something like that, saving the `<x>Form` naming system for the outermost `<form>` component. @radamson thoughts?

okeefm replied:
So this would go from `ResourceForm` to `ResourceElement` or `ResourceGroup` or the like

jason-crowley replied:
Yeah `ResourceForm` wasn't the best name, but I actually did away with that component in `json-xml-validation` because I created a new component called `FormInputItem` to represent the input-field-or-file-upload items (for both the resource and profile)

okeefm commented:
The only thing that might be useful is tests for the new components; other than that it looks good to me.

jason-crowley commented:
Alright, I added some tests, although they currently don't test much (since there's no validation to test in this PR). However the tests I added will pass on the `json-xml-validation` branch, which will have more comprehensive tests of the validation and interactions with the form

radamson commented on src/components/ProfileForm.tsx:
```diff
@@ -4,7 +4,7 @@ import { ProfileSelect } from './ProfileSelect';
 
 export interface ProfileProps {
-  readonly optionsByProfile: Map<String, SelectOption[]>;
+  readonly optionsByProfile: Map<string, SelectOption[]>;
```
is this actually a `Map`, or should it be a `Record`?
```suggestion
  readonly optionsByProfile: Record<string, SelectOption[]>;
```

jason-crowley replied:
Currently, it's a `Map`

radamson commented on src/__tests__/ValidatorForm.test.tsx:
```diff
@@ -0,0 +1,33 @@
+    render(<ValidatorForm basePath="" profiles={new Map()} />);
+  });
+
+  it.skip('displays the name of the file that was uploaded', async () => {
```
should these tests still be skipped?  If not we should make sure there are tickets to track implementing these missing features.

jason-crowley replied:
> However the tests that are skipped have been tested and will pass (after I push some changes) on the `json-xml-validation` branch, which will have more comprehensive tests of the validation and interactions with the form

Yes, they are skipped for now because I couldn't get it working with the jQuery in `app.js`, but we are no longer using `app.js` in `json-xml-validation`, and I got the tests to work on that branch.

radamson commented on src/components/ResourceForm.tsx:
```diff
@@ -0,0 +1,21 @@
+import React, { useState } from 'react';
+
+export function ResourceForm() {
+  const [input, setInput] = useState('');
```
we really should have types on these and the components, but we can save that for a future PR when the style checking is added.

radamson commented on src/components/ValidatorForm.tsx:
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+
+import { ResourceForm } from './ResourceForm';
+
```
unnecessary spacing
```suggestion
```

radamson commented on src/index.tsx:
```diff
@@ -1,10 +1,21 @@
 import ReactDOM from "react-dom";
 
+import { ValidatorForm } from './components/ValidatorForm';
+
```
unnecessary spacing

radamson commented on src/index.tsx:
```diff
@@ -1,10 +1,21 @@
+  const basePath = validatorElement.getAttribute('data-base-path');
+  const profiles: Map<string, string[]> = JSON.parse(atob(validatorElement.getAttribute('data-profiles')));
+  ReactDOM.render(
+    <React.StrictMode>
```
huh, TIL.  looks cool.

jason-crowley replied:
Hmm I can see from this change that I think you're right that it should be a `Record` after the `JSON.parse`

radamson replied:
I think you meant to comment on the other topic 🙂

jason-crowley replied:
I found out about StrictMode from the #react-js slack channel!

radamson commented on src/components/ValidatorForm.tsx:
```diff
@@ -0,0 +1,53 @@
+
+export interface ValidatorProps {
+  readonly basePath: string;
+  readonly profiles: Map<string, string[]>;
```
I think this should probably be a record

```suggestion
  readonly profiles: Record<string, string[]>;
```

radamson commented on src/components/ValidatorForm.tsx:
```diff
@@ -0,0 +1,53 @@
+import { SelectOption } from '../models/SelectOption';
+import { ProfileForm } from './ProfileForm';
+
+export interface ValidatorProps {
```
It doesn't look like this interface is used anywhere else so I don't think it needs to be exported.
